### DISCUSSION TITLE
Adding Hardware Support - Is there a guide?

### DISCUSSION THREAD

**Discussion by codecrete-ross** [Rem0o/FanControl.Releases#1623] [Q&A]
I'm a software developer and I'd like to add support for my motherboard (NZXT N7 Z370, honestly hate NZXT products now but here we are).

Is there a guide out there or any resources that help developers contribute to the hardware monitoring library?

**Comment by Rem0o** [comment]
You can check out this repo: https://github.com/LibreHardwareMonitor/LibreHardwareMonitor

There is already quite a bit of NZXT stuff in there.

**Reply by codecrete-ross** [reply]
@Rem0o I actually have the repo cloned and running in my IDE, but it's unclear to me how exactly I extend support.

Even though there is supported NZXT hardware in that application, only a few of my fans (e.g. fans plugged into the CPU, AIO, and Water Pump connectors) appear to work properly within Fan Control, so I was hoping for more guidance on how I can go about figuring out what's wrong and what to do to fix it. I would really love to be able to abandon the NZXT CAM application and use Fan Control as the main method of controlling my case fans.

On a completely unrelated note, the EVGA RTX 3080ti does work in Fan Control, but only two fans are controllable. The third appears to be governed by a secondary controller on the card, which does not appear in Fan Control, and it does not respond to settings applied to the two primary fans.

**Reply by Rem0o** [reply]
To extend support, you must have a bit of know-how as to how HID devices work, and must have access to some documentation as to how the protocol with that HID device work. Most of the time those protocols are reverse-engineered since manufacturers don't document them publicly.  You can take a look at NzxtGroup.cs, where it will identify HID device by a hard-coded ID for each device type.  That's where it starts.

Now as for your GPU, you should have 2 control cards in fan control, one per fan controller.   1 controller should control 2 fans.
